### PR TITLE
Emilce -Muncher (Consuming APIs)-Octos

### PR DESCRIPTION
**Pull request by emilcecarlisa**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? | I utilized the docs and postman.
Describe your API Wrapper.  How did you decide on the methods you created? | Class examples
Describe an edge case or failure case test you wrote for your API Wrapper. | 
Explain how VCR aids in testing an API. | They store test data so you're not making the same data calls to an API
What is the Heroku URL of your deployed application? | https://emilce-muncher.herokuapp.com/
Provide a link to the Trello board you used | https://trello.com/b/6ZtaMj2d/muncher

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Rails.application.routes.draw do
+	root 'recipes#search'
+
+	get '/recipes', to: 'recipes#index', as: 'recipes'
+	get '/recipe', to: 'recipes#show', as: 'recipe'
```
Comment (droberts-sea) [inline config/routes.rb:5]: Could you use `resources` for this?

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+
+    <title><%= content_for?(:title) ? yield(:title) : "Untitled" %></title>
+
```
Comment (droberts-sea) [inline app/views/layouts/application.html.erb:8]: This results in the page name being "Untitled"

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    <%= stylesheet_link_tag    "application" %>
+    <%= javascript_include_tag "application", 'data-turbolinks-track' => true %>
+    <%= csrf_meta_tags %>
+  </head>
+
+  <body>
+
+    <%= yield %>
+
```
Comment (droberts-sea) [inline app/views/layouts/application.html.erb:17]: You should have something here to help give your site a unified look and feel - maybe a `<header>` with the search bar, a `<footer>` with the Edamam logo, and `<main>` tag wrapped around the `yield`.

File: lib/recipe_api_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+		response = HTTParty.get(url)
+		recipe = initialize_recipe(response.first)
+		return recipe
+	end
+
+	private
+	def self.raise_on_error(response)
+		if response["OK"]
+			raise EdamamError.new(response["error"])
+		end
```
Comment (droberts-sea) [inline lib/recipe_api_wrapper.rb:42]: I like the idea of doing error handling in a method, but this doesn't match the way that Edamam reports errors. The `OK` and `errors` keys are what Slack did, but that's not going to be universal.

File: app/controllers/recipes_controller.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+class RecipesController < ApplicationController
+	MAX_PER_PAGE = 12
+
+	def index
+		@query = params[:q]
+		@from = params[:from].nil? ? 0 : params[:from].to_i
+		@to = params[:to].nil? ? @from + MAX_PER_PAGE : params[:to].to_i
+
+		@recipes = RecipeApiWrapper.list_recipes(@query, @from, @to)
+
```
Comment (droberts-sea) [inline app/controllers/recipes_controller.rb:12]: If you're expecting your lib file to raise an error, you should wrap this call in a `begin`/`rescue`.

File: test/lib/recipe_api_wrapper_test.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+require "test_helper"
+
+# Note You will also need to place the VCR.use_cassette block around your controller tests as well. Otherwise your controller tests, which use the API will fail.
+
+describe "Recipes" do
+	describe "list_recipes" do
+		it "lists recipes based on search" do
+			VCR.use_cassette("recipes") do
+				recipes = RecipeApiWrapper.list_recipes("ice cream")
```
Comment (droberts-sea) [inline test/lib/recipe_api_wrapper_test.rb:9]: What if a search yields no results?

File: test/lib/recipe_api_wrapper_test.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+				end
+			end
+		end
+	end
+
+	describe "get_details" do
+		it "returns the details for a specific recipe" do
+			VCR.use_cassette("recipes") do
+						search = "ice cream"
+						response = RecipeApiWrapper.get_details("http://www.edamam.com/ontologies/edamam.owl#recipe_7bf4a371c6884d809682a72808da7dc2")
```
Comment (droberts-sea) [inline test/lib/recipe_api_wrapper_test.rb:22]: What if you try to get details with a bogus URI?

File: test/controllers/recipes_controller_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe RecipesControllerController do
+  # it "must be a real test" do
+  #   flunk "Need real tests"
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_controller_test.rb:5]: You need some controller tests! Here is my list of what I'm looking for:

Index:
- Valid search term
- Search term that produces no results (e.g. `adfohdaglhjalj`)
- No search term

Show:
- Valid URI
- Bogus URI
- No URI

Root
- Does it work at all (only one test)

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | yes
Comprehension questions | some missing
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | yes
Semantic HTML | no
Errors are reported to the user | no - when I mangle the recipe URI, I get an exception rather than a polite error message
API Wrapper to handle the API requests | yes
Controller testing | no
Lib testing | missing failure cases, did not test the `Recipe` class 
Search Functionality | yes 
List Functionality | yes
Show individual item functionality (link to original recipe opens in new tab) | yes
**Styling** | 
Responsive layout | no
List View shows 10 items at a time/pagination | yes
The app is styled to create an attractive user interface | no
**API Features** | 
The App attributes Edaman | no - This is really important! When you're designing code for money, using someone else's work without attribution is a good way to end up in legal trouble.
The VCR casettes do not contain the API key | yes
**External Resources** | 
Link to Trello Board | yes
Link to deployed app on Heroku | yes
**Overall** |

This is a good start, but I think there's a way to go still. You were able to get the core API functionality working, which is good! However, there are some big holes still around error handling, testing and presentation. Remembering how to write "regular" Ruby code is part of the challenge of this assignment, but it seems like you're still struggling some of the Ruby fundamentals.

Many of the problems I'm seeing are things that will hamper the rest of your development process. For example, having a good story around error handling makes everything else easier, because when something goes wrong you get a clear reason why instead of having to hunt for it.

Fortunately, error handling and testing are big components of the VideoStore API project this week - make sure to spend some time focusing on this. These will continue to be important as we move into JavaScript next week, and having them down will make everything else you do a little easier.

I know you've been putting in a ton of effort recently, and I want to acknowledge that. I do see improvements happening and progress being made. Keep up your focus and dedication, and things will continue falling into place.
